src/lib.rs

```diff
@@ -27,11 +28,105 @@
 #[doc = include_str!("../README.md")]
 mod readme {}
 
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        let result = 2 + 2;
-        assert_eq!(result, 4);
+mod assert;
+mod format;
+mod utils;
+mod week;
+mod write;
+
+use core::fmt;
+
+/// Error type returned by the `strftime` functions.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Error {
+    /// Provided format string is ended by an unterminated format specifier.
+    InvalidFormatString,
+    /// Formatted string is too large and could cause an out-of-memory error.
+    FormattedStringTooLarge,
+    /// Provided buffer for the [`buffered::strftime`] function is too small for the formatted string.
+    ///
+    /// This corresponds to the [`std::io::ErrorKind::WriteZero`] variant.
+    WriteZero,
+    /// Formatting error, corresponding to [`core::fmt::Error`].
+    FmtError,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidFormatString => write!(f, "invalid format string"),
+            Error::FormattedStringTooLarge => write!(f, "formatted string too large"),
+            Error::WriteZero => write!(f, "failed to write the whole buffer"),
+            Error::FmtError => write!(f, "formatter error"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for Error {}
+
+/// Common methods needed for formatting _time_.
+pub trait Time {
+    /// Returns the year for _time_ (including the century).
+    fn year(&self) -> i32;
+    /// Returns the month of the year in `1..=12` for _time_.
+    fn month(&self) -> u8;
+    /// Returns the day of the month in `1..=31` for _time_.
+    fn day(&self) -> u8;
+    /// Returns the hour of the day in `0..=23` for _time_.
+    fn hour(&self) -> u8;
+    /// Returns the minute of the hour in `0..=59` for _time_.
+    fn minute(&self) -> u8;
+    /// Returns the second of the minute in `0..=60` for _time_.
+    fn second(&self) -> u8;
+    /// Returns the number of nanoseconds in `0..=999_999_999` for _time_.
+    fn nanoseconds(&self) -> u32;
+    /// Returns an integer representing the day of the week in `0..=6`, with `Sunday == 0`.
+    fn day_of_week(&self) -> u8;
+    /// Returns an integer representing the day of the year in `1..=366`.
+    fn day_of_year(&self) -> u16;
+    /// Returns the number of seconds as a signed integer since the Epoch.
+    fn to_int(&self) -> i64;
+    /// Returns true if the time zone is UTC.
+    fn is_utc(&self) -> bool;
+    /// Returns the offset in seconds between the timezone of _time_ and UTC.
+    fn utc_offset(&self) -> i32;
+    /// Returns the name of the time zone as a string.
+    fn time_zone(&self) -> &str;
+}
+
+pub mod buffered {
+    use super::{Error, Time};
+    use crate::format::TimeFormatter;
+
+    pub fn strftime(time: &impl Time, format: &[u8], mut buf: &mut [u8]) -> Result<(), Error> {
```

Comment:
I think we'll want the return type here to be `Result<&mut [u8], Error>`, kinda like the [`char::encode_utf8`](https://doc.rust-lang.org/stable/std/primitive.char.html#method.encode_utf8) so we return the subslice of the given `buf` we've written into.

That would make something like this work:

```rust
let mut buf = [0; 1000];
let time = todo!();
let out = strftime::buffered::strftime(&time, b"%Y", &mut buf)?;
assert_eq!(out, b"2022");
```

I think in order to make this easy, we'll want `TimeFormatter::fmt` to return `Result<usize, Error>`.